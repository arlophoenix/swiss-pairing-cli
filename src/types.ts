import { CLI_OPTION_FORMAT, CLI_OPTION_ORDER, SUPPORTED_FILE_TYPES } from './constants.js';

/**
 * Represents the options provided through the CLI
 */
export interface CLIOptions {
  readonly players?: readonly string[];
  readonly numRounds?: number;
  readonly startRound?: number;
  readonly matches?: readonly ReadonlyMatch[];
  readonly order?: CLIOptionOrder;
  readonly file?: string;
  readonly format?: CLIOptionFormat;
}

/**
 * Represents the order option for CLI
 */
export type CLIOptionOrder = (typeof CLI_OPTION_ORDER)[number];

/**
 * Represents the format option for CLI
 */
export type CLIOptionFormat = (typeof CLI_OPTION_FORMAT)[number];

/**
 * Represents the input for generating matches over multiple rounds
 */
export interface GenerateRoundMatchesInput {
  readonly players: readonly string[];
  readonly numRounds: number;
  readonly startRound: number;
  readonly playedOpponents: ReadonlyPlayedOpponents;
}

export type GenerateRoundMatchesOutputErrorType = 'InvalidInput' | 'NoValidSolution' | 'InvalidOutput';

/**
 * Represents the output for generating matches over multiple rounds
 */
export type GenerateRoundMatchesOutput =
  | { readonly success: true; readonly roundMatches: ReadonlyRoundMatches }
  | {
      readonly success: false;
      readonly errorType: GenerateRoundMatchesOutputErrorType;
      readonly errorMessage: string;
    };

export interface ValidateRoundMatchesInput {
  readonly players: readonly string[];
  readonly numRounds: number;
  readonly playedOpponents: ReadonlyPlayedOpponents;
}

export interface ValidateRoundMatchesOutput {
  readonly roundMatches: ReadonlyRoundMatches;
  readonly players: readonly string[];
  readonly numRounds: number;
  readonly playedOpponents: ReadonlyPlayedOpponents;
}

/**
 * Represents a generic result type
 */
export type Result<T> =
  | { readonly success: true; readonly value: T }
  | { readonly success: false; readonly errorMessage: string };

/**
 * Represents the result of a validation operation
 */
export type ValidationResult =
  | { readonly isValid: true }
  | { readonly isValid: false; readonly errorMessage: string };

/**
 * Represents the matches for multiple rounds
 */
export type RoundMatches = Record<string, readonly Match[]>;

/**
 * Represents the matches for multiple rounds (readonly version)
 */
export type ReadonlyRoundMatches = Record<string, readonly ReadonlyMatch[]>;

/**
 * Represents a match played between two players
 */
// eslint-disable-next-line functional/prefer-readonly-type
export type Match = [string, string];

/**
 * Represents a match played between two players (readonly version)
 */
export type ReadonlyMatch = readonly [string, string];

/**
 * Represents the opponents already played
 */
// eslint-disable-next-line functional/prefer-readonly-type
export type PlayedOpponents = Map<string, Set<string>>;

/**
 * Represents the opponents already played (readonly version)
 */
export type ReadonlyPlayedOpponents = ReadonlyMap<string, ReadonlySet<string>>;

export type SupportedFileTypes = (typeof SUPPORTED_FILE_TYPES)[number];
