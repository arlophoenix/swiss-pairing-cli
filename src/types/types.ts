export * from './errors.js';

import { ARGS, CLI_OPTION_FORMAT, CLI_OPTION_ORDER, SUPPORTED_FILE_TYPES } from '../constants.js';

export interface UnvalidatedCLIOptions {
  readonly players?: readonly string[];
  readonly numRounds?: string;
  readonly startRound?: string;
  readonly matches?: readonly (readonly string[])[];
  readonly order?: string;
  readonly file?: string;
  readonly format?: string;
}

export interface ValidatedCLIOptions {
  readonly players: readonly string[];
  readonly numRounds: number;
  readonly startRound: number;
  readonly matches: readonly ReadonlyMatch[];
  readonly order: CLIOptionOrder;
  readonly file: string;
  readonly format: CLIOptionFormat;
}

export type CLIOptionOrder = (typeof CLI_OPTION_ORDER)[number];
export type CLIOptionFormat = (typeof CLI_OPTION_FORMAT)[number];

export interface GenerateRoundMatchesInput {
  readonly players: readonly string[];
  readonly numRounds: number;
  readonly startRound: number;
  readonly playedOpponents: ReadonlyPlayedOpponents;
}

export interface ValidateRoundMatchesInput {
  readonly players: readonly string[];
  readonly numRounds: number;
  readonly playedOpponents: ReadonlyPlayedOpponents;
}

export interface ValidateRoundMatchesOutput {
  readonly roundMatches: ReadonlyRoundMatches;
  readonly players: readonly string[];
  readonly numRounds: number;
  readonly playedOpponents: ReadonlyPlayedOpponents;
}

export type RoundMatches = Record<string, readonly Match[]>;
export type ReadonlyRoundMatches = Record<string, readonly ReadonlyMatch[]>;

// eslint-disable-next-line functional/prefer-readonly-type
export type Match = [string, string];
export type ReadonlyMatch = readonly [string, string];
// eslint-disable-next-line functional/prefer-readonly-type
export type PlayedOpponents = Map<string, Set<string>>;
export type ReadonlyPlayedOpponents = ReadonlyMap<string, ReadonlySet<string>>;

export type SupportedFileTypes = (typeof SUPPORTED_FILE_TYPES)[number];

// export type CLIArg = keyof UnvalidatedCLIOptions;
export type CLIArg = (typeof ARGS)[number];
