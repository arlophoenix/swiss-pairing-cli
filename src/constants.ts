import { CLIOptionOrder } from './types.js';

export const ARG_MATCHES = 'matches';
export const ARG_NUM_ROUNDS = 'num-rounds';
export const ARG_ORDER = 'order';
export const ARG_PLAYERS = 'players';
export const ARG_START_ROUND = 'start-round';

export const BYE_PLAYER = 'BYE';

export const CLI_OPTION_ORDER = ['top-down', 'random', 'bottom-up'] as const;

export const CLI_OPTION_ORDER_DEFAULT: CLIOptionOrder = 'top-down';
