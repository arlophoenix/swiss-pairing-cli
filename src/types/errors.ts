export type Result<T> =
  | { readonly success: true; readonly value: T }
  | { readonly success: false; readonly error: ValidationError };

export type BooleanResult =
  | { readonly success: true }
  | { readonly success: false; readonly error: ValidationError };

export interface ValidationError {
  readonly type: ErrorType;
  readonly message: string;
}

export type ErrorType = 'InvalidInput' | 'InvalidOutput' | 'NoValidSolution';
