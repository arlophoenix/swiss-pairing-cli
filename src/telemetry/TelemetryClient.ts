import { AugmentedTelemetryEvent, TelemetryEvent } from './telemetryTypes.js';
import {
  detectEnvironment,
  detectExecutionContext,
  generateDistinctID,
  shouldEnableTelemetryClient,
} from './telemetryUtils.js';

import { Config } from '../Config.js';
import { DEBUG_TELEMETRY } from '../constants.js';
import { PostHog } from 'posthog-node';
import debug from 'debug';
import os from 'os';

const log = debug(DEBUG_TELEMETRY);

export class TelemetryClient {
  private readonly postHogClient: PostHog | null = null;
  private readonly distinctId!: string;
  private readonly enabled: boolean = true;
  // eslint-disable-next-line functional/prefer-readonly-type
  private eventQueue: AugmentedTelemetryEvent[] = [];
  // eslint-disable-next-line functional/prefer-readonly-type
  private flushTimeout: NodeJS.Timeout | null = null;

  // eslint-disable-next-line functional/prefer-readonly-type
  private static instance: TelemetryClient | null = null;

  public static getInstance(): TelemetryClient {
    if (!TelemetryClient.instance) {
      // eslint-disable-next-line functional/immutable-data
      TelemetryClient.instance = new TelemetryClient();

      // Ensure events are flushed on process exit
      process.on('exit', () => {
        TelemetryClient.instance?.shutdown().catch(() => {
          // nothing to do here the process has ended
        });
      });
    }
    return TelemetryClient.instance;
  }

  private constructor() {
    const config = Config.getInstance();
    const apiKey = config.getPosthogApiKey();
    this.enabled = shouldEnableTelemetryClient({
      telemetryOptOut: config.getTelemetryOptOut(),
      apiKeyExists: apiKey !== '',
      environment: detectEnvironment(),
    });

    log('Initializing telemetry');
    log('Telemetry enabled:', this.enabled);

    if (this.enabled) {
      try {
        this.postHogClient = new PostHog(apiKey, {
          host: 'https://us.i.posthog.com',
          flushAt: 1,
          flushInterval: 0,
        });
        this.distinctId = generateDistinctID();
      } catch (error) {
        log('Failed to initialize telemetry client', error);
        this.enabled = false;
      }
    }
  }

  record(event: TelemetryEvent) {
    if (!this.enabled || !this.postHogClient) {
      return;
    }
    log('Recording event:', event);
    const augmentedEvent: AugmentedTelemetryEvent = {
      name: event.name,
      properties: {
        ...event.properties,
        node_version: process.version,
        os_name: os.platform(),
        cli_version: process.env.npm_package_version,
        execution_context: detectExecutionContext(),
        environment: detectEnvironment(),
      },
    };

    // Queue the event
    // eslint-disable-next-line functional/immutable-data
    this.eventQueue.push(augmentedEvent);

    // Debounce flush
    if (this.flushTimeout) {
      clearTimeout(this.flushTimeout);
    }
    // eslint-disable-next-line functional/immutable-data
    this.flushTimeout = setTimeout(() => {
      void this.flush();
    }, 100);
  }

  private async flush(): Promise<void> {
    if (this.postHogClient === null || this.eventQueue.length === 0) {
      return;
    }
    log('Flushing %d events', this.eventQueue.length);
    try {
      // Send all queued events in parallel
      await Promise.all(
        this.eventQueue.map((event) => {
          this.postHogClient?.capture({
            distinctId: this.distinctId,
            event: event.name,
            properties: event.properties,
          });
        })
      );
      // eslint-disable-next-line functional/immutable-data
      this.eventQueue = [];
    } catch {
      // Silently fail
    } finally {
      log('Flush complete');
    }
  }

  async shutdown(): Promise<void> {
    if (this.flushTimeout) {
      clearTimeout(this.flushTimeout);
    }

    // Final attempt to flush any queued events
    await this.flush();

    if (this.postHogClient) {
      try {
        await this.postHogClient.shutdown();
      } catch {
        // Ignore shutdown errors
      }
    }
  }

  public static resetForTesting() {
    // eslint-disable-next-line functional/immutable-data
    this.instance = null;
  }
}
