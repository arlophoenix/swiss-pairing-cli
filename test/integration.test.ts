import { afterEach, beforeEach, describe, expect, it, test } from '@jest/globals';

import { SUPPORTED_FILE_TYPES } from '../src/constants.js';
import { SupportedFileTypes } from '../src/types/types.js';
import { TelemetryClient } from '../src/telemetry/TelemetryClient.js';
import { exec } from 'child_process';
import fs from 'fs';
import fsPromise from 'fs/promises';
import os from 'os';
import path from 'path';
import util from 'util';

const execAsync = util.promisify(exec);
const fixturesDir = path.join(__dirname, 'fixtures');

const fileCache: Record<string, string> = {};
const cliResultCache = new Map<string, CLIResult>();

interface ExecError extends Error {
  readonly stdout: string;
  readonly stderr: string;
  readonly code: number;
}

interface CLIResult {
  readonly success: boolean;
  readonly message: string;
}

function isExecError(error: unknown): error is ExecError {
  return (
    typeof error === 'object' && error !== null && 'stderr' in error && 'stdout' in error && 'code' in error
  );
}

async function readFileContent(filePath: string): Promise<string> {
  if (!fileCache[filePath]) {
    // eslint-disable-next-line functional/immutable-data
    fileCache[filePath] = await fsPromise.readFile(filePath, 'utf-8');
  }
  return fileCache[filePath];
}

async function runCLI({
  args,
  env = process.env,
  useCache = false,
}: {
  readonly args: string;
  readonly env?: NodeJS.ProcessEnv;
  readonly useCache?: boolean;
}): Promise<CLIResult> {
  const cacheKey = `args:${args}`;
  const cliResult = cliResultCache.get(cacheKey);
  if (useCache && cliResult) {
    return cliResult;
  }

  try {
    const { stdout, stderr } = await execAsync(`node dist/index.js ${args}`, { env });
    const success = stderr === '';
    const result = { success: success, message: success ? stdout : `Error: ${stderr}\n${stdout}` };
    if (useCache) {
      cliResultCache.set(cacheKey, result);
    }
    return result;
  } catch (error) {
    if (isExecError(error)) {
      const result = { success: false, message: error.stderr };
      if (useCache) {
        cliResultCache.set(cacheKey, result);
      }
      return result;
    }
    throw error;
  }
}

function validateCLIResult({
  result,
  isErrorCase,
}: {
  readonly result: CLIResult;
  readonly isErrorCase: boolean;
}): void {
  // print the output for unexpected errors
  if (result.success === isErrorCase) {
    console.log(result.message);
  }
  expect(result.success).toBe(!isErrorCase);
  expect(result).toMatchSnapshot();
}

async function validateFixture({
  fileName,
  isErrorCase,
  allFixtures,
}: {
  readonly fileName: string;
  readonly isErrorCase: boolean;
  readonly allFixtures: readonly string[];
}): Promise<void> {
  const ext = path.extname(fileName);
  const fixturePath = path.join(fixturesDir, fileName);

  // text files contain arguments to be run directly in the CLI
  if (ext === '.txt') {
    const input = await readFileContent(fixturePath);
    const result = await runCLI({ args: input, useCache: true });
    validateCLIResult({ result, isErrorCase });
    // JSON or CSV files are expected to be provided as a file argument
  } else if (SUPPORTED_FILE_TYPES.includes(ext as SupportedFileTypes)) {
    const result = await runCLI({ args: `--file ${fixturePath}`, useCache: true });
    validateCLIResult({ result, isErrorCase });

    // Compare file input with direct CLI args if both exist
    const correspondingTxtFixtureName = fileName.replace(ext, '.txt');
    if (allFixtures.includes(correspondingTxtFixtureName)) {
      const correspondingTxtFixturePath = path.join(fixturesDir, correspondingTxtFixtureName);
      const inputWithArgs = await readFileContent(correspondingTxtFixturePath);
      const resultWithArgs = await runCLI({ args: inputWithArgs, useCache: true });
      expect(result).toEqual(resultWithArgs);
    }
  } else {
    throw new Error(`Unsupported file type: ${ext}`);
  }
}

describe('Integration Tests', () => {
  // Filter for only our test files, excluding hidden files and system files
  const allFixtures = fs
    .readdirSync(fixturesDir)
    .filter(
      (file) =>
        !file.startsWith('.') &&
        (SUPPORTED_FILE_TYPES.includes(path.extname(file) as SupportedFileTypes) ||
          path.extname(file) === '.txt')
    );
  const successFixtures = allFixtures.filter((file) => !file.startsWith('invalid-'));
  const failureFixtures = allFixtures.filter((file) => file.startsWith('invalid-'));

  test.each(successFixtures)('CLI Output: Success Case - %s', async (fixtureName) => {
    await validateFixture({ fileName: fixtureName, isErrorCase: false, allFixtures });
  });
  test.each(failureFixtures)('CLI Output: Failure Case - %s', async (fixtureName) => {
    await validateFixture({ fileName: fixtureName, isErrorCase: true, allFixtures });
  });

  describe('Telemetry Notice Integration', () => {
    let configDir: string;
    const originalEnv = { ...process.env };

    beforeEach(() => {
      // Set up temp directory for config files
      configDir = path.join(os.tmpdir(), `swiss-pairing-test-${String(Date.now())}`);
    });

    afterEach(() => {
      // eslint-disable-next-line functional/immutable-data
      process.env = originalEnv;
      if (fs.existsSync(configDir)) {
        fs.rmSync(configDir, { recursive: true });
      }
      TelemetryClient.resetForTesting();
    });

    it('should show notice on first run and create notice file', async () => {
      const env = {
        ...process.env,
        XDG_CONFIG_HOME: configDir,
        SWISS_PAIRING_TELEMETRY_OPT_OUT: '',
      };
      // First run
      const result1 = await runCLI({
        args: '--teams Alice Bob',
        env,
      });
      expect(result1.success).toBe(true);
      expect(result1.message).toContain('Telemetry Notice');

      // Second run
      const result2 = await runCLI({
        args: '--teams Alice Bob',
        env,
      });
      expect(result2.success).toBe(true);
      expect(result2.message).not.toContain('Telemetry Notice');
    });

    it('should respect telemetry opt out', async () => {
      const env = {
        ...process.env,
        XDG_CONFIG_HOME: configDir,
        SWISS_PAIRING_TELEMETRY_OPT_OUT: '1',
      };
      const result = await runCLI({ args: '--teams Alice Bob', env });
      expect(result.success).toBe(true);
      expect(result.message).not.toContain('Telemetry Notice');
    });
  });
});
