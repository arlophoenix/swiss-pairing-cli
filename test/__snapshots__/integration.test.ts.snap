// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Tests CLI Output - duplicate-match-in-history.txt 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team3
2. Team2 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - five-teams-three-rounds-start-two-bottom-up-format-text-with-matches.json 1`] = `
{
  "message": "# Matches

**Round 3**

1. BYE vs Elvis
2. David vs Bob
3. Charlie vs Alice

**Round 4**

1. BYE vs David
2. Elvis vs Alice
3. Charlie vs Bob
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams.csv 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team2
2. Team3 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams.json 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team2
2. Team3 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams.txt 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team2
2. Team3 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-bottom-up.csv 1`] = `
{
  "message": "**Round 1**

1. Team4 vs Team3
2. Team2 vs Team1
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-bottom-up.json 1`] = `
{
  "message": "**Round 1**

1. Team4 vs Team3
2. Team2 vs Team1
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-bottom-up.txt 1`] = `
{
  "message": "**Round 1**

1. Team4 vs Team3
2. Team2 vs Team1
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-top-down.csv 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team2
2. Team3 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-top-down.json 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team2
2. Team3 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-top-down.txt 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team2
2. Team3 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-with-matches-bottom-up.csv 1`] = `
{
  "message": "**Round 1**

1. Team4 vs Team2
2. Team3 vs Team1
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-with-matches-bottom-up.json 1`] = `
{
  "message": "**Round 1**

1. Team4 vs Team2
2. Team3 vs Team1
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-with-matches-bottom-up.txt 1`] = `
{
  "message": "**Round 1**

1. Team4 vs Team2
2. Team3 vs Team1
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-with-matches-top-down.csv 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team3
2. Team2 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-with-matches-top-down.json 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team3
2. Team2 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-with-matches-top-down.txt 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team3
2. Team2 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-with-squads.csv 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team3
2. Team2 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-with-squads.json 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team3
2. Team2 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - four-teams-with-squads.txt 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team3
2. Team2 vs Team4
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R6-1-unmodified.txt 1`] = `
{
  "message": "**Round 6**

1. Morgath vs Pauldog
2. Nzdaemon vs GreyGhost
3. Saravock vs lilme
4. Arlo vs Colain
5. armydan vs LCYig
6. iso.evil vs Clin
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R6-1-unmodified-bottom-up.csv 1`] = `
{
  "message": "**Round 6**

1. LCYig vs armydan
2. Clin vs iso.evil
3. Colain vs Arlo
4. GreyGhost vs Nzdaemon
5. Pauldog vs Morgath
6. lilme vs Saravock
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R6-1-unmodified-bottom-up.json 1`] = `
{
  "message": "**Round 6**

1. LCYig vs armydan
2. Clin vs iso.evil
3. Colain vs Arlo
4. GreyGhost vs Nzdaemon
5. Pauldog vs Morgath
6. lilme vs Saravock
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R6-1-unmodified-bottom-up.txt 1`] = `
{
  "message": "**Round 6**

1. LCYig vs armydan
2. Clin vs iso.evil
3. Colain vs Arlo
4. GreyGhost vs Nzdaemon
5. Pauldog vs Morgath
6. lilme vs Saravock
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R6-2-win-gg.txt 1`] = `
{
  "message": "**Round 6**

1. Morgath vs Pauldog
2. Nzdaemon vs GreyGhost
3. Saravock vs lilme
4. Arlo vs Colain
5. armydan vs LCYig
6. iso.evil vs Clin
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R6-3-win-ad.txt 1`] = `
{
  "message": "**Round 6**

1. Morgath vs armydan
2. Nzdaemon vs Pauldog
3. Saravock vs lilme
4. Arlo vs LCYig
5. GreyGhost vs Colain
6. iso.evil vs Clin
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R6-4-draw.txt 1`] = `
{
  "message": "**Round 6**

1. Morgath vs Pauldog
2. Nzdaemon vs GreyGhost
3. Saravock vs lilme
4. Arlo vs Colain
5. armydan vs LCYig
6. iso.evil vs Clin
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R7.txt 1`] = `
{
  "message": "**Round 7**

1. Morgath vs armydan
2. Nzdaemon vs Pauldog
3. Saravock vs Colain
4. lilme vs iso.evil
5. Arlo vs LCYig
6. GreyGhost vs Clin
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R8-1-unmodified.txt 1`] = `
{
  "message": "**Round 8**

1. Nzdaemon vs armydan
2. Morgath vs Colain
3. Saravock vs LCYig
4. Arlo vs iso.evil
5. lilme vs GreyGhost
6. Pauldog vs Clin
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R8-1-unmodified-team-names.csv 1`] = `
{
  "message": "**Round 8**

1. Howl-Terrroa vs The High Life
2. Morgath Marauders vs The Good Runts
3. Kabal of Khaine vs Lyzyrd Skyzyrd
4. Under Where? vs Mean & Green
5. Friendly Unit vs Dark Accolytes
6. The Barrowtown Brawlers vs AttackAttack
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R8-1-unmodified-team-names.txt 1`] = `
{
  "message": "**Round 8**

1. Howl-Terrroa vs The High Life
2. Morgath Marauders vs The Good Runts
3. Kabal of Khaine vs Lyzyrd Skyzyrd
4. Under Where? vs Mean & Green
5. Friendly Unit vs Dark Accolytes
6. The Barrowtown Brawlers vs AttackAttack
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R8-2-m-win-or-draw.txt 1`] = `
{
  "message": "**Round 8**

1. Morgath vs LCYig
2. Nzdaemon vs Colain
3. Saravock vs GreyGhost
4. Arlo vs armydan
5. lilme vs iso.evil
6. Pauldog vs Clin
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R8-2-m-win-or-draw-team-names.csv 1`] = `
{
  "message": "**Round 8**

1. Morgath Marauders vs Mean & Green
2. Howl-Terrroa vs The Good Runts
3. Kabal of Khaine vs Lyzyrd Skyzyrd
4. Under Where? vs The High Life
5. Friendly Unit vs Dark Accolytes
6. The Barrowtown Brawlers vs AttackAttack
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - hawbbl24S2R8-2-m-win-or-draw-team-names.txt 1`] = `
{
  "message": "**Round 8**

1. Morgath Marauders vs Mean & Green
2. Howl-Terrroa vs The Good Runts
3. Kabal of Khaine vs Lyzyrd Skyzyrd
4. Under Where? vs The High Life
5. Friendly Unit vs Dark Accolytes
6. The Barrowtown Brawlers vs AttackAttack
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - invalid-duplicate-teams.txt 1`] = `
{
  "message": "Invalid CLI argument "--teams": "Team1,Team2,Team1,Team3". Expected unique team names
",
  "success": false,
}
`;

exports[`Integration Tests CLI Output - invalid-match-format.txt 1`] = `
{
  "message": "Invalid CLI argument "--matches": "["Team1",null]". Expected an array of two team names
",
  "success": false,
}
`;

exports[`Integration Tests CLI Output - invalid-num-rounds.txt 1`] = `
{
  "message": "Invalid input: Number of rounds (4) must be less than number of teams (4)
",
  "success": false,
}
`;

exports[`Integration Tests CLI Output - invalid-played-self.txt 1`] = `
{
  "message": "Invalid input: Team1 cannot play against itself
",
  "success": false,
}
`;

exports[`Integration Tests CLI Output - invalid-single-team.txt 1`] = `
{
  "message": "Invalid CLI argument "--teams": "Team1". Expected at least two teams
",
  "success": false,
}
`;

exports[`Integration Tests CLI Output - invalid-squad-format.txt 1`] = `
{
  "message": "Invalid CLI argument "--teams": "Team1 [A". Expected valid team name, optionally followed by [squad] e.g."Alice [Home]"
",
  "success": false,
}
`;

exports[`Integration Tests CLI Output - invalid-start-round.txt 1`] = `
{
  "message": "Invalid CLI argument "--start-round": "0". Expected a positive integer
",
  "success": false,
}
`;

exports[`Integration Tests CLI Output - three-teams-bottom-up.csv 1`] = `
{
  "message": "**Round 1**

1. BYE vs Team3
2. Team2 vs Team1
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - three-teams-bottom-up.json 1`] = `
{
  "message": "**Round 1**

1. BYE vs Team3
2. Team2 vs Team1
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - three-teams-bottom-up.txt 1`] = `
{
  "message": "**Round 1**

1. BYE vs Team3
2. Team2 vs Team1
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - three-teams-top-down.csv 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team2
2. Team3 vs BYE
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - three-teams-top-down.json 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team2
2. Team3 vs BYE
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - three-teams-top-down.txt 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team2
2. Team3 vs BYE
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - three-teams-with-matches-bottom-up.csv 1`] = `
{
  "message": "**Round 1**

1. BYE vs Team2
2. Team3 vs Team1
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - three-teams-with-matches-bottom-up.json 1`] = `
{
  "message": "**Round 1**

1. BYE vs Team2
2. Team3 vs Team1
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - three-teams-with-matches-bottom-up.txt 1`] = `
{
  "message": "**Round 1**

1. BYE vs Team2
2. Team3 vs Team1
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - three-teams-with-matches-top-down.csv 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team3
2. Team2 vs BYE
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - three-teams-with-matches-top-down.json 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team3
2. Team2 vs BYE
",
  "success": true,
}
`;

exports[`Integration Tests CLI Output - three-teams-with-matches-top-down.txt 1`] = `
{
  "message": "**Round 1**

1. Team1 vs Team3
2. Team2 vs BYE
",
  "success": true,
}
`;
